ocaml/xapi/xapi_cluster.ml

```diff
@@ -47,6 +47,50 @@ let create ~__context ~pIF ~cluster_stack ~pool_auto_join ~token_timeout ~token_
       assert_pif_prerequisites (pIF,pifrec);
       let ip = ip_of_pif (pIF,pifrec) in
 
+      let assert_pool_on_same_subnet () =
+        (** we gather ip addresses from all the hosts and also the netmask from the PIF.
+          * if all the ip addresses do not have the same network address, then we raise an
+            error *)
+        let are_on_same_subnet =
+            let parse_ip ip_str =
+              match Ipaddr.of_string ip_str with
+              | Ok ip_addr       -> ip_addr
+              (* this shouldn't happen because we shouldn't be storing invalid ips *)
+              | Error (`Msg msg) -> raise Api_errors.(Server_error (internal_error, [Printf.sprintf "could not parse ip address (%s), msg: %s" ip_str msg]))
+            in
+            let pool_host_ips =
+               Xapi_pool_helpers.get_master_slaves_list ~__context |>
+               List.map (fun x -> Db.Host.get_address ~__context ~self:x) |>
+               List.map parse_ip
+            in
+            let ensure_same_protocol ips =
+              let netmask = pifrec.API.pIF_netmask |> parse_ip in
+              let coerce_v4 xs = xs |> List.map (function
+                | Ipaddr.V4 x -> x
+                | V6 _        -> raise Api_errors.(Server_error (internal_error, ["expected IPv4, but got IPv6"])))
+              in
+              let coerce_v6 xs = xs |> List.map (function
+                | Ipaddr.V6 x -> x
+                | V4 _        -> raise Api_errors.(Server_error (internal_error, ["expected IPv6, but got IPv4"])))
+              in
+              match netmask with
+              | Ipaddr.V4 nm -> `Ip_v4s (nm, coerce_v4 ips)
+              | V6        nm -> `Ip_v6s (nm, coerce_v6 ips)
+            in
+            match ensure_same_protocol pool_host_ips with
+            | `Ip_v4s (netmask, ip::ips) ->
+              let network_addr = Ipaddr.V4.Prefix.of_netmask netmask ip in
+              List.for_all (fun x -> Ipaddr.V4.Prefix.mem x network_addr) ips
+            | `Ip_v6s (netmask, ip::ips) ->
+              let network_addr = Ipaddr.V6.Prefix.of_netmask netmask ip in
+              List.for_all (fun x -> Ipaddr.V6.Prefix.mem x network_addr) ips
+            | `Ip_v4s (_, []) | `Ip_v6s (_, []) -> raise Api_errors.(Server_error (internal_error, ["impossible: pool should contain at least one member"]))
+        in
+        if not are_on_same_subnet then
+          raise Api_errors.(Server_error (internal_error, ["not every host is on the same subnet"]))
```

Comment:
If the error is due to something that the user has done wrong, then it is not an internal error. Instead, define a new API error in the datamodel and api_errors.ml.